cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Setup.java

```diff
@@ -0,0 +1,137 @@
+
+        // compute w (base used in double spending protection, see 2020 incsys paper)
+        GroupElement w = (GroupElement) hf.hash("w" + bg.toString()); // TODO: ... .toString ok? in PR anmerken
+
+        // compute seventh base for user token (h7 in 2020 incsys paper)
+        GroupElement h7 = (GroupElement) hf.hash("h7" + bg.toString());
+
+        // instantiate prf used in this instance of the incentive system with a proper key length
+        AesPseudorandomFunction prf = new AesPseudorandomFunction(PRF_KEY_LENGTH);
+
+        // instantiate SPS-EQ scheme used in this instance of the incentive system
+        SPSEQSignatureScheme spsEq = new SPSEQSignatureScheme(new SPSEQPublicParameters(bg));
+        SignatureKeyPair<SPSEQVerificationKey, SPSEQSigningKey> spsEqKeyPair = spsEq.generateKeyPair(2);
+        // TODO: save key pair in public parameters
+
+        // wrap up all values
+        return new PublicParameters(bg, w, h7, prf, spsEq, spsEqKeyPair);
+    }
+
+    /**
+     * generates a user key pair from public parameters
+     * @param pp object representation of public parameters
+     * @return object representation of a user key pair (see 2020 inc sys paper)
+     */
+    public static UserKeyPair userKeyGen(PublicParameters pp)
+    {
+        // draw random exponent for the user secret key
+        Zn usedZn = pp.getBG().getZn(); // the remainder class ring used in this instance of the incentive system
+        ZnElement usk = usedZn.getUniformlyRandomElement(); // secret exponent
+
+        // generate the user key for the PRF used to generate pseudorandomness in the system
+        PrfKey betaUsr = pp.getPrf().generateKey(); // key the user uses to generate pseudorandomness
+
+        // compute user public key from secret exponent
+        GroupElement upkElem = pp.getW().pow(usk);
+
+        // wrap up values
+        UserSecretKey USK = new UserSecretKey(usk, betaUsr);
+        UserPublicKey upk = new UserPublicKey(upkElem);
+        return new UserKeyPair(upk, USK);
+    }
+
+
+    public static ProviderKeyPair providerKeyGen(PublicParameters pp)
+    {
+        // draw the dlogs of the first 6 bases used in the Pedersen commitment in the token
+        ZnElement[] q = new ZnElement[6];
+        for(int i=0; i < q.length; i++)
+        {
+            q[i] = pp.getBG().getZn().getUniformlyRandomElement();
+        }
+
+        // compute above first 6 bases
+        GroupElement[] h = new GroupElement[6];
+        GroupElement g1Generator = pp.getBG().getG1().getGenerator();
+        for(int i = 0; i < h.length; i++)
+        {
+            h[i] =g1Generator.pow(q[i]);
+        }
+
+        // generate PRF key for provider
+        PrfKey betaProv = pp.getPrf().generateKey();
+
+        // wrap up values
+
+        SPSEQVerificationKey spseqVerificationKey = pp.getSpsEq().getVerificationKey(
+                pp.getSpsEqKeyPair().getVerificationKey().getRepresentation()
```

Comment:
Generally there's never a need to serialize, then deserialize something within the same statement. 
Just use 
`SPSEQVerificationKey spseqVerificationKey = pp.getSpsEqKeyPair().getVerificationKey()`
(but obviously don't use that because pp shouldn't contain these keys anyway - just for the sake of illustration that the `Representation` intermediate step is unnecessary).

Reply:
As far as I see a version taking no parameters does not exist for SPSEQSignatureScheme.getVerificationKey()